### PR TITLE
[NVPTX] Support !"cluster_dim_{x,y,z}" metadata

### PR DESCRIPTION
**Pull request by AlexMaclean**
Add support for !"cluster_dim_{x,y,z}" metadata to allow specifying cluster dimensions on a kernel function in llvm.

If any of these metadata entries are present, the `.explicitcluster` PTX directive is used and the specified dimensions are lowered with the `.reqnctapercluster` directive. For more details see: [PTX ISA: 11.7. Cluster Dimension Directives](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#cluster-dimension-directives)

### CODE REVIEW

File: llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp [hunk 1]
```diff
@@ -563,21 +563,40 @@ void NVPTXAsmPrinter::emitKernelFunctionDirectives(const Function &F,
...
+  if (STI->getSmVersion() >= 90) {
+    std::optional<unsigned> ClusterX = getClusterDimx(F);
+    std::optional<unsigned> ClusterY = getClusterDimy(F);
+    std::optional<unsigned> ClusterZ = getClusterDimz(F);
+
+    if (ClusterX || ClusterY || ClusterZ) {
+      O << ".explicitcluster\n";
+      if (ClusterX.value_or(1) != 0) {
+        assert(ClusterY.value_or(1) && ClusterZ.value_or(1) &&
+               "clusterx != 0 implies clustery and clusterz should be non-zero "
```
Comment (durga4github) [inline llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp:586]: In the assert messages, can we have the x/y/z suffixes in upper-case?
For example, "clustery" to: clusterY or cluster_dim_y or clusterDimY or something like that..

File: llvm/test/CodeGen/NVPTX/cluster-dim.ll [hunk 2]
```diff
@@ -0,0 +1,17 @@
+; RUN: llc < %s -march=nvptx -mcpu=sm_90 | FileCheck %s
+; RUN: llc < %s -march=nvptx64 -mcpu=sm_90 | FileCheck %s
+; RUN: %if ptxas %{ llc < %s -march=nvptx -mcpu=sm_90 | %ptxas-verify %}
+; RUN: %if ptxas %{ llc < %s -march=nvptx64 -mcpu=sm_90 | %ptxas-verify %}
```
Comment (durga4github) [inline llvm/test/CodeGen/NVPTX/cluster-dim.ll:4]: Do we need "ptxas-12.0" here, since this is only supported from sm_90 onwards?

Comment (AlexMaclean) [reply llvm/test/CodeGen/NVPTX/cluster-dim.ll]: good point, fixed

File: llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp (hunk 1 above)
Comment (AlexMaclean) [reply llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp]: I've renamed to "cluster_dim_{x,y,z}"

File: llvm/test/CodeGen/NVPTX/cluster-dim.ll (hunk 2 above)
Comment (durga4github) [reply llvm/test/CodeGen/NVPTX/cluster-dim.ll]: okay, thanks!

**Review by durga4github — COMMENTED** [review]
Changes looks good to me.

File: llvm/test/CodeGen/NVPTX/cluster-dim.ll [hunk 3]
```diff
@@ -0,0 +1,15 @@
+; RUN: llc < %s -march=nvptx64 -mcpu=sm_90 | FileCheck %s
+; RUN: %if ptxas-12.0 %{ llc < %s -march=nvptx64 -mcpu=sm_90 | %ptxas-verify -arch=sm_90 %}
```
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/cluster-dim.ll:2]: I'd also add a run with pre-sm90 GPU, and switch to auto-generated checks, so it captures the "no-attributes here" case for older GPUs.

File: llvm/lib/Target/NVPTX/NVPTXUtilities.cpp [hunk 4]
```diff
@@ -130,8 +130,8 @@ static void cacheAnnotationFromMD(const Module *m, const GlobalValue *gv) {
   }
 }
 
-bool findOneNVVMAnnotation(const GlobalValue *gv, const std::string &prop,
-                           unsigned &retval) {
+std::optional<unsigned> findOneNVVMAnnotation(const GlobalValue *gv,
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXUtilities.cpp:133]: Nice. Thank you for updating the old code to use better modern C++ features.

**Review by Artem-B — COMMENTED** [review]
LGTM in principle.

Would you be willing to land it as a separate commit? It's a nice self-contained improvement that's not directly related to the cluster_dim changes.

File: llvm/test/CodeGen/NVPTX/cluster-dim.ll (hunk 3 above)
Comment (AlexMaclean) [reply llvm/test/CodeGen/NVPTX/cluster-dim.ll:4]: Sounds good. Updated.

**Review by Artem-B — APPROVED** [review]

**Pull request merged**